simphony_mayavi/cuds/vtk_lattice.py

```diff
@@ -121,52 +155,99 @@ def iter_nodes(self, indices=None):
+
+        Returns
+        -------
+        coordinates : tuple of float[3]
+        '''
         point_id = self._get_point_id(index)
         return self.data_set.get_point(point_id)
 
     # Alternative constructors ###############################################
 
     @classmethod
-    def empty(cls, name, type_, base_vector, size, origin, data=None):
+    def empty(cls, name, primitive_cell, size, origin, data=None):
         """ Create a new empty Lattice.
 
         """
-        if type_ in ('Cubic', 'OrthorombicP', 'Square', 'Rectangular'):
-            data_set = tvtk.ImageData(spacing=base_vector, origin=origin)
+        bravais_lattice = primitive_cell.bravais_lattice
+        if bravais_lattice in (BravaisLattice.CUBIC, BravaisLattice.TETRAGONAL,
+                               BravaisLattice.ORTHORHOMBIC):
+            # Compute the spacing from the primitive cell
+            spacing = tuple(vector_len(p) for p in (primitive_cell.p1,
+                                                    primitive_cell.p2,
+                                                    primitive_cell.p3))
+            data_set = tvtk.ImageData(spacing=spacing, origin=origin)
             data_set.extent = 0, size[0] - 1, 0, size[1] - 1, 0, size[2] - 1
-        elif type_ == 'Hexagonal':
-            x, y, z = numpy.meshgrid(
-                range(size[0]), range(size[1]), range(size[2]))
+        elif bravais_lattice in BravaisLattice:
+            y, z, x = numpy.meshgrid(
+                range(size[1]), range(size[2]), range(size[0]))
             points = numpy.zeros(shape=(x.size, 3), dtype='double')
-            points[:, 0] += base_vector[0] * x.ravel() \
-                + 0.5 * base_vector[0] * y.ravel()
-            points[:, 1] += base_vector[1] * y.ravel()
-            points[:, 0] += origin[0]
-            points[:, 1] += origin[1]
-            points[:, 2] += origin[2]
+            # construct points using primitive cells
+            for idim in range(3):
+                points[:, idim] += primitive_cell.p1[idim]*x.ravel() +\
+                    primitive_cell.p2[idim]*y.ravel() +\
+                    primitive_cell.p3[idim]*z.ravel()
+                points[:, idim] += origin[idim]
             data_set = tvtk.PolyData(points=points)
         else:
-            message = 'Unknown lattice type: {}'.format(type_)
-            raise ValueError(message)
-        return cls(name=name, type_=type_, data=data, data_set=data_set)
+            message = 'Unknown lattice type: {}'
+            raise ValueError(message.format(str(bravais_lattice)))
+
+        return cls(name=name, primitive_cell=primitive_cell,
+                   data=data, data_set=data_set)
 
     @classmethod
     def from_lattice(cls, lattice):
         """ Create a new Lattice from the provided one.
 
         """
-        base_vectors = lattice.base_vect
+        pc = lattice.primitive_cell
         origin = lattice.origin
-        lattice_type = lattice.type
```

Comment:
`suggestion`: you can still set `lattice_type = primitive_cell.bravais_lattice` or  probably better `bravais_lattice=primitive_cell.bravais_lattice`. I think it will make the code more readable